File: mobile/lib/features/bipintar/quiz/quiz_config.dart
import 'dart:math';
import '../../../core/constants/sign_labels.dart';

/// Konfigurasi & aturan skor untuk mode Quiz "Tantangan Isyarat" (Camera Challenge).
class QuizConfig {
  /// Jumlah soal per sesi.
  static const int questionsPerSession = 10;

  /// Waktu maksimal per soal (detik).
  static const int timePerQuestionSec = 30;

  /// Confidence minimum agar deteksi dihitung sebagai jawaban benar.
  static const double matchScore = 0.5;

  /// Lama (ms) isyarat benar harus ditahan stabil sebelum dihitung benar.
  static const int holdMs = 700;

  /// Interval loop game internal (ms).
  static const int tickMs = 100;

  static const int basePoints = 50;
  static const int timeBonusPerSec = 2;
  static const int streakBonus = 10;
}

/// Hitung poin satu soal benar.
int computeQuestionScore(double timeLeftSec, int streakAfter) {
  final timeBonus = max(0, timeLeftSec.floor()) * QuizConfig.timeBonusPerSec;
  final streakBonusVal = max(0, streakAfter - 1) * QuizConfig.streakBonus;
  return QuizConfig.basePoints + timeBonus + streakBonusVal;
}

/// Acak label lalu ambil sejumlah soal (Fisher–Yates via shuffle).
List<SignLabel> buildQuestionQueue(List<SignLabel> labels, int count) {
  final pool = List<SignLabel>.from(labels)..shuffle();
  if (count <= pool.length) return pool.sublist(0, count);
  final result = <SignLabel>[];
  while (result.length < count) {
    result.addAll(pool);
  }
  return result.sublist(0, count);
}

File: web/src/lib/quizConfig.ts
import { SignLabel } from "@/constants/signLabels";

/**
 * Konfigurasi & aturan skor untuk mode Quiz "Tantangan Isyarat" (Camera Challenge).
 * Dipakai bersama oleh hook game (useSignQuiz) dan halaman quiz.
 */
export const QUIZ_CONFIG = {
  /** Jumlah soal per sesi permainan */
  questionsPerSession: 10,
  /** Waktu maksimal per soal (detik) */
  timePerQuestionSec: 30,
  /**
   * Confidence minimum agar sebuah deteksi dihitung sebagai jawaban.
   * Lebih tinggi dari CONFIDENCE_THRESHOLD deteksi biasa (0.25) agar tidak asal lolos.
   */
  matchScore: 0.5,
  /**
   * Berapa lama (ms) isyarat yang benar harus DITAHAN stabil sebelum dihitung benar.
   * Mencegah false-positive akibat kedipan deteksi satu-dua frame.
   */
  holdMs: 700,
  /** Interval loop game internal (ms) */
  tickMs: 100,
} as const;

/** Poin dasar untuk setiap jawaban benar */
export const BASE_POINTS = 50;
/** Poin bonus per detik sisa waktu */
export const TIME_BONUS_PER_SEC = 2;
/** Poin bonus per level streak (beruntun) */
export const STREAK_BONUS = 10;

/**
 * Hitung poin satu soal yang dijawab benar.
 * @param timeLeftSec sisa waktu (detik) saat benar
 * @param streakAfter nilai streak SETELAH jawaban benar ini (>=1)
 */
export function computeQuestionScore(timeLeftSec: number, streakAfter: number): number {
  const timeBonus = Math.max(0, Math.floor(timeLeftSec)) * TIME_BONUS_PER_SEC;
  const streakBonus = Math.max(0, streakAfter - 1) * STREAK_BONUS;
  return BASE_POINTS + timeBonus + streakBonus;
}

/**
 * Acak urutan label lalu ambil sejumlah soal (tanpa pengulangan bila cukup).
 * Fisher–Yates shuffle.
 */
export function buildQuestionQueue(labels: SignLabel[], count: number): SignLabel[] {
  const pool = [...labels];
  for (let i = pool.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [pool[i], pool[j]] = [pool[j], pool[i]];
  }
  if (count <= pool.length) return pool.slice(0, count);
  // Bila diminta lebih banyak dari jumlah label, ulangi pool teracak.
  const result: SignLabel[] = [];
  while (result.length < count) {
    result.push(...pool);
  }
  return result.slice(0, count);
}
